File: AlgoExpert/Arrays/12-Longest-Peak/Java/Solution.java

class Solution {
  public static int longestPeak(int[] array) {
    // Write your code here.
    if (array.length < 3) {
      return 0;
    }

    int prev = array[0];
    int longestPeak = 0;
    int currPeak = 1;
    boolean inDecline = false;

    for (int i = 1; i < array.length; i++) {
      int currNumber = array[i];

      if (prev == currNumber) {
        currPeak = 1;
      } else if (currNumber > prev && !inDecline) {
        currPeak += 1;
      } else if (currNumber > prev && inDecline) {
        inDecline = false;
        currPeak = 1;
      } else if (currNumber < prev && currPeak >= 2 && !inDecline) {
        inDecline = true;
        currPeak += 1;
        longestPeak = Math.max(longestPeak, currPeak);
      } else if (currNumber < prev && inDecline) {
        currPeak += 1;
        longestPeak = Math.max(longestPeak, currPeak);
      } else if (currNumber < prev && currPeak < 2) {
        currPeak = 1;
      }

      prev = currNumber;
    }

    return longestPeak;
  }
}

File: AlgoExpert/Arrays/12-Longest-Peak/Python/solution.py
def longestPeak(array):
  # Write your code here.
  if len(array) < 3:
    return 0
  
  prev = array[0]
  longest_peak = 0
  curr_peak = 1
  in_decline = False

  for i in range(1, len(array)):
    curr_number = array[i]

    if prev == curr_number:
      curr_peak = 1
    elif curr_number > prev and not in_decline:
      curr_peak += 1
    elif curr_number > prev and in_decline:
      in_decline = False
      curr_peak = 2
    elif curr_number < prev and not in_decline and curr_peak != 1:
      in_decline = True
      curr_peak += 1
      longest_peak = max(longest_peak, curr_peak)
    elif curr_number < prev and in_decline and curr_peak != 1:
      curr_peak += 1
      longest_peak = max(longest_peak, curr_peak)

    prev = curr_number

  return longest_peak